iup/__init__.py

```diff
@@ -996,112 +1010,42 @@ def predict(
 
 
 #### prediction output ####
-class Forecast(ValidateData):
+class QuantileForecast(ValidatedData):
     """
-    Abstract class for all the forecast data type.
+    Class for forecast with quantiles.
+    Save for future.
     """
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-    # Can add validate to enforce the specific schema of Forecast when more Forecast types are added
+    # Must be named as "quantileXX" except the date column
     def validate(self):
+        # has at least 1 column of date and 1 column of estimate
         self.assert_type_included(pl.Date)
         self.assert_type_included(pl.Float64)
-        # has at least 1 column of date and 1 column of estimate
-
-    def assert_type_included(self, datatype):
-        """
-        Verify at least one column has the expected type.
-
-        Parameters
-        columns: List[str]
-            names of columns for which to check type
-        datatype:
-            data type for each listed column
-        """
-        assert any(
-            dtype == datatype for dtype in self.schema.values()
-        ), f"No column is {datatype} type"
-
-
-class QuantileForecast(Forecast):
-    """
-    Class for forecast with quantiles.
-    Save for future.
-    """
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-    # Must be named as "quantileX.XX" except the date column
-    def validate(self):
+        # except date, must have the common column names
         estimate = self.select(pl.all().exclude(pl.Date))
-        QuantileForecast.assert_column_name_all(estimate, "quantile")
-
-    @staticmethod
-    def assert_column_name_all(estimate, column_name):
-        """
-        Verify that all columns have a pattern with a common name
-
-        Parameters
-        estimate:
-            The data without 'date' type
-        column_name:
-            The common column name
-        """
-        pattern = rf"^{column_name}\d(\.\d+)?$"
-        assert all(
-            [bool(re.match(pattern, col)) for col in estimate.columns]
-        ), f"Not all columns are Column name {column_name}"
+        super().assert_column_name_all(estimate, "quantile")
 
 
 class PointForecast(QuantileForecast):
     """
     Class for forecast with point estimate
-    A subclass when quantile is 0.5
-    For now, enforce the "quantile0.5" to be "estimate"
+    A subclass when quantile is 50%
+    For now, enforce the "quantile50" to be "estimate"
     """
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     def validate(self):
-        self.assert_columns_found(["estimate"])
-        self.assert_columns_type(["estimate"], pl.Float64)
+        super().assert_columns_found(["estimate"])
```

Comment:
Having a bank of `assert_something` methods at the level of the parent class and then choosing the right combination of these methods to validate each child class seems clever to me. Nice work, Fuhan!